geochem_inverse_optimize.py

```diff
@@ -75,56 +75,148 @@ def get_sample_graphs(data_dir: str) -> Tuple[nx.DiGraph, Any]:
+        self.sample_network = sample_network
+        self.sample_adjacency = sample_adjacency
+        self._site_to_parameter: Dict[str, cp.Parameter] = {}
+        self._primary_terms = []
+        self._regularizer_terms = []
+        self._regularizer_strength = cp.Parameter(nonneg=True)
+        self._problem = None
+        self._build_primary_terms()
+        self._build_regularizer_terms()
+        self._build_problem()
+
+    def _build_primary_terms(self) -> None:
+        for _, data in self.sample_network.nodes(data=True):
+            data["data"].total_flux = 0.0
+
+        # Build the main objective
+        # Use a topological sort to ensure an upstream-to-downstream traversal
+        for sample_name, my_data in nx_topological_sort_with_data(self.sample_network):
+            # Set up a CVXPY parameter for each element for each node
+            my_data.my_value = cp.Variable(pos=True)
+
+            # area weighted contribution from this node
+            # TODO(rbarnes): Try scaling area of a region by dividing by total area of all regions. `my_data.normalized_area`
+            my_data.my_flux = my_data.area * my_data.my_value
+
+            # Add the flux I generate to the total flux passing through me
+            my_data.total_flux += my_data.my_flux
+
+            # Normalise observation by mean
+            observed = cp.Parameter(pos=True)
+            self._site_to_parameter[my_data.data.name] = observed
+            normalised_concentration = my_data.total_flux / my_data.total_area
+            self._primary_terms.append(cp_log_ratio_norm(normalised_concentration, observed))
+
+            if ds := nx_get_downstream(self.sample_network, sample_name):
+                downstream_data = self.sample_network.nodes[ds]["data"]
+                # Add our flux to the downstream node's
+                downstream_data.total_flux += my_data.total_flux
+
+    def _build_regularizer_terms(self) -> None:
+        # Build the regularizer
+        sample_names = [
+            node for node in self.sample_network.nodes
+        ]  # Sample names for each node. TODO. MAKE THIS A DICTIONARY PART OF THE CLASS
+        for adjacent_nodes, border_length in self.sample_adjacency.items():
+            # NB adjacency network stores samples as numbers from 1 - 63 (*not* 0 - 62)
+            node_a = sample_names[adjacent_nodes[0] - 1]  # Get samplename from node 'number'
+            node_b = sample_names[adjacent_nodes[1] - 1]
+            a_concen = self.sample_network.nodes[node_a]["data"].my_value
+            b_concen = self.sample_network.nodes[node_b]["data"].my_value
+            #           self._regularizer_terms.append(border_length * (cp_log_ratio_norm(a_concen,b_concen)))  # < 'log-ratio' difference (preferable)
+            self._regularizer_terms.append(
+                border_length * (a_concen - b_concen)
+            )  # simple difference (not desirable)
```

Comment:
Note that after formatting the comment ends up in an awkward place. Please move comment about Line 131 and make it `TODO`.